repos/system_upgrade/common/libraries/overlaygen.py

```diff
@@ -42,39 +137,74 @@ def _mount_name(mountpoint):
 def _mount_dir(mounts_dir, mountpoint):
     return os.path.join(mounts_dir, _mount_name(mountpoint))
 
+def _get_mountpoint(mount_points, dir_path):
+    for mp in sorted(mount_points, reverse=True):
+        mod_mp = mp if mp[-1] == '/' else '{}/'.format(mp)
+        if dir_path.startswith(mod_mp):
+            # longest first, so the first one we find, is the last mp on the path
+            return mp
 
-def _prepare_required_mounts(scratch_dir, mounts_dir, mount_points, xfs_info):
-    result = {
-        mount_point.fs_file: mounting.NullMount(
-            _mount_dir(mounts_dir, mount_point.fs_file)) for mount_point in mount_points
-    }
-
-    if not xfs_info.mountpoints_without_ftype:
-        return result
 
-    space_needed = _overlay_disk_size() * len(xfs_info.mountpoints_without_ftype)
+def _prepare_required_mounts(scratch_dir, mounts_dir, storage_info, scratch_reserve):
+    """
+    Create disk images and loop mount them.
+
+    Ensure to create disk image for each important mountpoint configured
+    in fstab (excluding fs types noted in `OVERLAY_DO_NOT_MOUNT`).
+    Disk images reflect the free space of related partition/volume. In case
+    of partition hosting /var/lib/leapp/* calculate the free space value
+    taking `scratch_reserve` into account, as during the run of the tooling,
+    we will be consuming the space on the partition and we want to be more
+    sure that we do not consume all the space on the partition during the
+    execution - so we reduce the risk we affect run of other applications
+    due to missing space.
+
+    Note: the partition hosting the scratch dir is expected to be the same
+    partition that is hosting the target userspace container, but it does not
+    have to be true if the code changes. Right now, let's live with that.
+
+    See `_create_mount_disk_image` docstring for additional more details.
+
+    :param scratch_dir: Path to the scratch directory.
+    :type scratch_dir: str
+    :param mounts_dir: Path to the directory supposed to be a mountpoint.
+    :type mounts_dir: str
+    :param storage_info: The StorageInfo message.
+    :type storage_info: leapp.models.StorageInfo
+    :param scratch_reserve: Number of MB that should be extra reserved in a partition hosting the scratch_dir.
+    :type scratch_reserve: Optional[int]
+    """
+    mount_points = _get_mountpoints(storage_info)
+    scratch_mp = _get_mountpoint(mount_points, scratch_dir)
     disk_images_directory = os.path.join(scratch_dir, 'diskimages')
 
     # Ensure we cleanup old disk images before we check for space constraints.
+    # NOTE(pstodulk): Could we improve the process so we create imgs & calculate
+    # the required disk space just once during each leapp (pre)upgrade run?
     run(['rm', '-rf', disk_images_directory])
```

Comment:
What about adding some exception handling here?

Reply:
Thinking about that and around since the morning. Will see

Reply:
let's keep it for now as it is. We have similar action on multiple places.